code/modules/paperwork/contract.dm

```diff
@@ -0,0 +1,293 @@
+	else
+		return ..()
+
+/obj/item/weapon/paper/contract/infernal/attack(mob/M, mob/living/user)
+	add_fingerprint(user)
+	if(M == user && target == M.mind && M.mind.soulOwner != owner && attempt_signature(user, 1))
+		user.visible_message("<span class='danger'>[user] slices their wrist with [src], and scrawls their name in blood.</span>", "<span class='danger'>You slice your wrist open and scrawl your name in blood.</span>")
+		if(istype(user, /mob/living/carbon/human))
+			var/mob/living/carbon/human/C = user
+			C.vessel.remove_reagent("blood",100)
+	else
+		return ..()
+
+/obj/item/weapon/paper/contract/infernal/proc/attempt_signature(mob/living/carbon/human/user, blood = 0)
+	if(user.IsAdvancedToolUser() && user.is_literate())
+		if(user.mind == target)
+			if(user.mind.soulOwner != owner)
+				if (contractType == CONTRACT_REVIVE)
+					to_chat(user,"<span class='notice'>You are already alive, this contract would do nothing.</span>")
+				else
+					if(signed)
+						to_chat(user,"<span class='notice'>This contract has already been signed.  It may not be signed again.</span>")
+					else
+						to_chat(user,"<span class='notice'>You quickly scrawl your name on the contract</span>")
+						if(FulfillContract(target.current, blood)<=0)
+							to_chat(user,"<span class='notice'>But it seemed to have no effect, perhaps even Hell itself cannot grant this boon?</span>")
+						return 1
+			else
+				to_chat(user, "<span class='notice'>This devil already owns your soul, you may not sell it to them again.</span>")
+		else
+			to_chat(user,"<span class='notice'>Your signature simply slides off the sheet, it seems this contract is not meant for you to sign.</span>")
+	else
+		to_chat(user, "<span class='notice'>You don't know how to read or write.</span>")
+	return 0
+
+
+
+/obj/item/weapon/paper/contract/infernal/revive/attack(mob/M, mob/living/user)
+	if (target == M.mind && M.stat == DEAD && M.mind.soulOwner == M.mind)
+		var/mob/living/carbon/human/H = M
+		var/mob/dead/observer/ghost = H.get_ghost()
+		var/response = "No"
+		if(ghost)
+			ghost.notify_cloning("A devil has offered you revival, at the cost of your soul.",'sound/effects/genetics.ogg', H)
+			response = alert(ghost, "A devil is offering you another chance at life, at the price of your soul, do you accept?", "Infernal Resurrection", "Yes", "No")
+			if(!ghost)
+				return		//handle logouts that happen whilst the alert is waiting for a response.
+		else
+			response = alert(target.current, "A devil is offering you another chance at life, at the price of your soul, do you accept?", "Infernal Resurrection", "Yes", "No")
+		if(response == "Yes")
+			H.revive(1,0)
+			add_logs(user, H, "infernally revived via contract")
+			user.visible_message("<span class='notice'>With a sudden blaze, [H] stands back up.</span>")
+			H.fakefire()
+			FulfillContract(H, 1)//Revival contracts are always signed in blood
+			spawn(5)
+				H.fakefireextinguish(TRUE)
+	else
+		..()
```

Comment:
Perhaps the devil should get a notification if the target accepts, denies, or lacks a ghost. Also, not sure how this will work out if the ghost is still in their body